KAFKA-5130: Refactor TC In-memory Cache

guozhangwang opened the pull request:
1. Collapsed the `ownedPartitions`, `pendingTxnMap` and the `transactionMetadataCache` into a single in-memory structure, which is a two-layered map: first keyed by the transactionTxnLog, and then valued with the current coordinatorEpoch of that map plus another map keyed by the transactional id.

2. Use `transactionalId` across the modules in transactional coordinator, attach this id with the transactional marker entries.

3. Use two keys: `transactionalId` and `txnLogPartitionId` in the writeMarkerPurgatory as well as passing it along with the TxnMarkerEntry, so that `TransactionMarkerRequestCompletionHandler` can use it to access the two-layered map upon getting responses.

4. Use one queue per `broker-id` and `txnLogPartitionId`. Also when there is a possible update on the end point associated with the `broker-id`, update the Node without clearing the queue but relying on the requests to retry in the next round.

5. Centralize the error handling callback for appending-metadata-to-log and sending-markers-to-brokers in `TransactionStateManager#appendTransactionToLog`, and `TransactionMarkerChannelManager#addTxnMarkersToSend`.

6. Always update the in-memory transaction metadata AFTER the txn log has been appended and replicated, and then double check on the cache to make sure nothing has changed since log appending. The only exception is when initializing the pid for the first time, in which we will put a dummy into the cache but set its pendingState as `Empty` (so it will be valid to transit from `Empty` to `Empty`) so that it can be updated after the log append has completed.

guozhangwang committed:
collapse ownedPartition with MetadataCache

guozhangwang committed:
centralize callbacks for send markers and append to log

guozhangwang committed:
minor refactoring

guozhangwang committed:
rebase from trunk

guozhangwang committed:
minor fixes

guozhangwang commented:
@dguy @junrao 

I am still working to fixing all the unit tests but the non-testing code is ready for reviews.

apurvam commented:
Looks like the build fails with compilation errors..

guozhangwang commented:
@apurvam those are unit test compilations. I'm still working on those but non-testing code does build.

apurvam commented:
ah.. ok.. thanks!

ijuma commented on core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerChannelManager.scala:
```diff
@@ -98,55 +99,91 @@ object TransactionMarkerChannelManager {
+                s"has been appended to the log. The partition ${txnStateManager.partitionFor(transactionalId)} may have migrated as the metadata is no longer in the cache")
+          }
+
+        case errors: Errors =>
```
There's no need to have the `Errors` type parameter here. That check always succeeds.

guozhangwang replied:
ack

ijuma commented on core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala:
```diff
@@ -89,19 +88,37 @@ class TransactionStateManager(brokerId: Int,
+          case null => None
+          case txnMetadata => Some(CoordinatorEpochAndTxnMetadata(txnMetadataCacheEntry.coordinatorEpoch, txnMetadata))
+        }
+      case None => None
```
This can be written more clearly as:

```scala
transactionMetadataCache.get(partitionId).flatMap { cacheEntry =>
  cacheEntry.metadataPerTransactionalId.get(transactionalId)
}.map { txnMetadata =>
  CoordinatorEpochAndTxnMetadata(txnMetadataCacheEntry.coordinatorEpoch, txnMetadata)
}
```

guozhangwang replied:
ack. this needs to be tweaked a bit since the inner map is a `Pool`, let me know if you like the new pattern or not.

ijuma commented on core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala:
```diff
@@ -60,22 +62,19 @@ class TransactionStateManager(brokerId: Int,
   this.logIdent = "[Transaction Log Manager " + brokerId + "]: "
 
-  type WriteTxnMarkers = WriteTxnMarkerArgs => Unit
+  type SendTxnMarkersCallback = (String, Int, TransactionMetadata, TransactionResult) => Unit
```
Hmm, as discussed previously, it's preferable to use a case class instead of many unnamed parameters.

guozhangwang replied:
EDIT: the reason I made it as unnamed parameters is that the `sendTxnMarkers` function is not long used as a parameter but also directly called elsewhere. And it has been reduced from 8 params to 4, so I feel this is better cost-effective?

dguy commented on core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala:
```diff
@@ -194,21 +212,23 @@ class TransactionCoordinator(brokerId: Int,
+            val metadata = epochAndMetadata.transactionMetadata
+
+            if (metadata.producerId != pid) {
               (Errors.INVALID_PID_MAPPING, null)
```
Do we need to update these tuples to be `(error, null, null)`? i.e., we are expecting a triple as the result

guozhangwang replied:
ack.

dguy commented on core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala:
```diff
@@ -411,9 +393,8 @@ class TransactionCoordinator(brokerId: Int,
 
 case class InitPidResult(pid: Long, epoch: Short, error: Errors)
+
 case class WriteTxnMarkerArgs(transactionalId: String,
```
Looks like this is not used anymore?

guozhangwang replied:
ack

dguy commented on core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerChannel.scala:
```diff
@@ -84,102 +77,70 @@ class TransactionMarkerChannel(interBrokerListenerName: ListenerName,
+        }
       }
-      brokerId.get
+      brokerNode.get
```
Why do we need to do this?

guozhangwang replied:
this is needed as `groupBy` to generate the grouped map keyed by the node object.

dguy replied:
yeah - duh! I missed that!

dguy commented on core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerChannelManager.scala:
```diff
@@ -98,55 +99,91 @@ object TransactionMarkerChannelManager {
+                        // retry appending
+                        txnStateManager.appendTransactionToLog(transactionalId, coordinatorEpoch, txnMetadata, retryAppendCallback)
+
+                      case errors: Errors =>
```
Should we not just retry in this case? We've already written `PrepareXX` to the log and i thought we need to complete the transaction? How will the transaction be completed?

guozhangwang replied:
From `appendTransactionToLog` the only error codes we would pass in the callback are:

```
COORDINATOR_NOT_AVAILABLE
NOT_COORDINATOR
UNKNOWN
other
```

And anything falling into `other` should be considered fatal as they should not happen (if they happen that should be a bug), and `UNKNOWN` comes from `MESSAGE_TOO_LARGE` or `RECORD_LIST_TOO_LARGE` which are also doomed as fatal.

dguy replied:
ok - that makes sense. However, if this does happen how would we recover from it? We'd have uncommitted data in the logs which would block consumers using read committed. Do we have any plans to deal with this?

guozhangwang replied:
When this happens usually ops people need to be involved to manually truncate the log before starting. But I think this code can also be improved a bit for a more graceful shutdown.

Do you have any ideas?

dguy commented on core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerChannelManager.scala:
```diff
@@ -98,55 +99,91 @@ object TransactionMarkerChannelManager {
+
+            case None =>
+              // this transactional id no longer exists, maybe the corresponding partition has already been migrated out.
+              // return NOT_COORDINATOR to let the client re-discover the transaction coordinator
```
The comment is incorrect. I don't think we are returning anything to the client. If the transaction has been emmigrated than it the commit should be completed by the new partition leader

guozhangwang replied:
ack. updating the comment.

dguy commented on core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerChannelManager.scala:
```diff
@@ -98,55 +99,91 @@ object TransactionMarkerChannelManager {
+          }
+
+        case errors: Errors =>
+          throw new IllegalStateException(s"Unexpected error ${errors.exceptionName} before appending to txn log for $transactionalId")
```
I think we also need a check `case Errors.NOT_COORDINATOR` in which case we should just log an move on. 
Also, in this case i think we need to retry?

guozhangwang replied:
From `DelayedTxnMarker` when we call `completionCallback` we will only return `NONE` and hence any other error codes would not be expected and considered as fatal right? Note that the write-marker-response could contain error codes like `UNKNOWN_TOPIC_OR_PARTITION`, but this is handled in the `TransactionMarkerRequestCompletionHandler`, in which case we will retry.

dguy commented on core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerRequestCompletionHandler.scala:
```diff
@@ -52,43 +52,50 @@ class TransactionMarkerRequestCompletionHandler(transactionMarkerChannel: Transa
+                    retryPartitions += topicPartition
+
+                  case errors: Errors =>
+                    throw new IllegalStateException(s"Unexpected error ${errors.exceptionName} while sending txn marker for $transactionalId")
```
How are we cleaning up cases like this where it has failed? What are the implications for consumers etc?

guozhangwang replied:
Generally this default is just for any un-expected error codes (i.e. there is a bug in the protocol, that broker returns some error codes not defined) to fail-fast.

dguy commented on core/src/main/scala/kafka/coordinator/transaction/TransactionMetadata.scala:
```diff
@@ -138,37 +138,141 @@ private[coordinator] class TransactionMetadata(val pid: Long,
+
+            false
+          } else {
+            txnStartTimestamp = newMetadata.txnStartTimestamp
```
why do we need to update the `txnStartTimestamp` here and in the `CompleteCommit` case?

guozhangwang replied:
I have not reviewed your expiration PR so my understanding might be inconsistent, but here is how I read the existing code: `txnStartTimestamp` is only set when there is an active txn ongoing, otherwise it should always be set to -1 (i.e. the above check on the new metadata). And expiration will simply skip the metadata when this field is -1.

dguy replied:
My understanding is that it only needs to be set when we add the first partition to a new transaction.

guozhangwang replied:
When a transaction is completed, should we reset the startTimestamp to -1 again to avoid being expired?

junrao replied:
Yes, it seems Damian's understanding makes sense.

dguy commented on core/src/main/scala/kafka/coordinator/transaction/TransactionMetadata.scala:
```diff
@@ -138,37 +138,141 @@ private[coordinator] class TransactionMetadata(val pid: Long,
+            true
+          }
+
+        case PrepareAbort => // from endTxn
```
could this be `case PrepareAbort | PrepareCommit` ? I think they are identical

guozhangwang replied:
good point, ack.

dguy commented on core/src/main/scala/kafka/coordinator/transaction/TransactionMetadata.scala:
```diff
@@ -138,37 +138,141 @@ private[coordinator] class TransactionMetadata(val pid: Long,
+            true
+          }
+
+        case CompleteAbort => // from write markers
```
`case CompleteAbort | Complete Commit` ? I think they are identical

guozhangwang replied:
good point, ack.

dguy commented on core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala:
```diff
@@ -271,31 +272,22 @@ class TransactionStateManager(brokerId: Int,
-      ownedPartitions.remove(partition)
-      loadingPartitions.remove(partition)
-    }
+    val topicPartition = new TopicPartition(Topic.TransactionStateTopicName, partitionId)
```
We have the situation here that when this method returns nothing has actually happened yet, i.e., the removal is done on a different thread. So what happens if we get a request at the same time for another transactionalId that is in the same partition? Are we just relying/hoping it will eventually fail?

guozhangwang replied:
This will be in completed in another PR for the locking mechanism, and here is a sketch:

1. we will add a per-broker read-write lock on TC.
2. All request-handling logic need to grab the read lock before accessing the cache for checking state until appending to the local log call returns.
3. the loading / removing thread will grab the write lock when removing / adding the sub-map for that partition into the cache.
4. implementation-wise, we will modify `stateLock` for that read-write lock. The main purpose is to avoid log appending out-of-order, e.g. consider the following order:

```
a. check on cache passed.
b. write to local txn log

between a and b, long GC happens and emigration happens, followed by an immigration so that the broker re-become the tc for that txn partition. So step b) will succeed but there could be more entries already appended after checking step a)
```

dguy replied:
In step 2 isn't the log append callback happening on another thread? So how do we release the read lock? I guess i'm missing something but i'm reading it like this:
```
inReadLock(lock) {
   // do some stuff
   txnManager.appendToLog(...)
}
```
Which would mean the read lock would be given up as soon as the `appendToLog` call returns, but that doesn't mean the write to the log has actually completed as it might go into purgatory and be completed by another thread. I guess you have some other idea?

guozhangwang replied:
As I said we only need to make sure the the log entry is appended to the local data segment (even not required to flushed on disk), not necessarily replicated complete.

dguy replied:
Ok  - i missed the local bit. Thanks

dguy commented on core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala:
```diff
@@ -271,31 +272,22 @@ class TransactionStateManager(brokerId: Int,
+            info(s"Removed ${txnMetadataCacheEntry.metadataPerTransactionalId.size} cached transaction metadata for $topicPartition on follower transition")
+
+          case None =>
+            throw new IllegalStateException(s"The metadata cache entry for txn partition $partitionId does not exist while trying to remove it.")
```
Is this an exceptional condition? Could we just log it and move on?

guozhangwang replied:
Hmm, that is a good point. I think it can happen that `onFollower` called twice but the second call's triggered thread gets executed before the first call. So yeah we could just log it and move on.

guozhangwang replied:
EDIT: similarly for `addLoadedTransactionsToCache`, but slightly different: we only have one background thread inside `scheduler` and we are checking on `loadingPartitions` so that no requests will be handled while loading so no new entries will be appended to the log. So even we have two loading tasks scheduled,  it is safe for the second loaded metadata to replace the first loaded sub-map via `put`, so if `isDefined` we can just log it.

dguy commented on core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala:
```diff
@@ -310,8 +302,9 @@ class TransactionStateManager(brokerId: Int,
 
   // TODO: check broker message format and error if < V2
   def appendTransactionToLog(transactionalId: String,
+                             coordinatorEpoch: Int,
```
Should we change this to have `CoordinatorEpochAndTxnMetadata` as a param rather than the 2 separate params?

guozhangwang replied:
I have been thinking about it, the main reason that I keep it separately is that it can be called in two paths: handler thread call it directly and the callback from writeMarkerSender call it separately. In the latter case the two values are separate so I ended up keep them as is.

guozhangwang commented on core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala:
```diff
@@ -60,22 +62,19 @@ class TransactionStateManager(brokerId: Int,
   this.logIdent = "[Transaction Log Manager " + brokerId + "]: "
 
-  type WriteTxnMarkers = WriteTxnMarkerArgs => Unit
+  type SendTxnMarkersCallback = (String, Int, TransactionMetadata, TransactionResult) => Unit
```
ack. rebasing again..

guozhangwang committed:
address github comments

hachikuji commented on core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala:
```diff
@@ -89,19 +88,35 @@ class TransactionStateManager(brokerId: Int,
+  def getTransactionState(transactionalId: String): Option[CoordinatorEpochAndTxnMetadata] = {
+    val partitionId = partitionFor(transactionalId)
+
+    transactionMetadataCache.get(partitionId).flatMap { cacheEntry =>
```
Why is it safe to access `transactionMetadataCache` without a lock? It seems we protect mutations with the state lock, but since it is not a concurrent collection, don't we need to protect reads as well?

guozhangwang replied:
1. Please see my previous comment: we will modify the `statelock` into a read-write lock and the only operations that could mutate the top-level map `transactionMetadataCache` will be covered (the checking / append-to-log will be covered in read lock).

2. For the lower-level map `Pool[String, TransactionMetadata]`, its operations are thread safe and any modifications to the inner `TransactionMetadata` will be covered by the object `synchronized` itself and all modifications will be in-place than object override.

hachikuji replied:
Sounds good. Will that be part of this patch?

hachikuji commented on core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala:
```diff
@@ -40,8 +41,9 @@ import scala.collection.JavaConverters._
 
 object TransactionManager {
```
Not from this patch, but should this be `TransactionStateManager`?

guozhangwang replied:
ack.

hachikuji commented on core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala:
```diff
@@ -89,19 +88,35 @@ class TransactionStateManager(brokerId: Int,
+          CoordinatorEpochAndTxnMetadata(txnMetadataCacheEntry.coordinatorEpoch, txnMetadata)
+        }
+
+     case None =>
```
nit: this is not aligned.

guozhangwang replied:
ack.

hachikuji commented on core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala:
```diff
@@ -109,73 +109,91 @@ class TransactionCoordinator(brokerId: Int,
-            metadata synchronized {
-              if (!metadata.eq(newMetadata))
-                initPidWithExistingMetadata(transactionalId, transactionTimeoutMs, responseCallback, metadata)
+            epochAndMetadata synchronized {
```
Hmm... It doesn't seem right to synchronize on the instance of `CoordinatorEpochAndTxnMetadata`. We construct a new one in every call to `addTransaction`, right? Also, we're still synchronizing on the metadata elsewhere in this file.

guozhangwang replied:
Yeah that is a good question, ideally we should sycnrhonize on `txnMetadata` only, since coordinator epoch will only likely be changed during loading / removal of the metadata. Will refactor on this part.

hachikuji commented on core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala:
```diff
@@ -109,73 +109,91 @@ class TransactionCoordinator(brokerId: Int,
+
+    if (metadata.state == Ongoing) {
+      // abort the ongoing transaction and then return CONCURRENT_TRANSACTIONS to let client wait and retry
+      def callback(errors: Errors): Unit = {
```
nit: `error`? In spite of the name of `Errors`, there is only one error. A few more of these in this class if you're keen.

guozhangwang replied:
ack.

hachikuji commented on core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala:
```diff
@@ -109,73 +109,91 @@ class TransactionCoordinator(brokerId: Int,
+
+    // set the pending state to Empty again to allow the metadata to be updated
+    // after it was successfully written to log
+    metadata.prepareTransitionTo(Empty)
```
This is a comment throughout, but it seems we rarely check the result of `prepareTransitionTo`. Perhaps that function should just raise an invalid state exception if the state transition is invalid? Otherwise we should get in the habit of adding the checks.

guozhangwang replied:
We check it in `TransactionMetadata`, that the new state is equal to the pending state.

hachikuji replied:
But if we know the transition is invalid here, then we could skip appending to the log, right? I feel we should just make `prepareTransitionTo` raise an exception if the attempted transition is invalid.

guozhangwang replied:
EDIT: discussed offline, realized it is a different issue than I thought about. Refactored this part a bit as well.

hachikuji commented on core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala:
```diff
@@ -109,73 +109,91 @@ class TransactionCoordinator(brokerId: Int,
-                             metadata: TransactionMetadata,
-                             initPidCallback: InitPidCallback): Unit ={
+  private def initPidWithNewMetadata(transactionalId: String,
+                                     initPidCallback: InitPidCallback,
```
nit: Not from this patch, but should this be `responseCallback`? It's nice to keep naming consistent and we use this name in `handleInitPid` and `initPidWithExistingMetadata`. It's also helpful that the name expresses that this callback returns to the user.

guozhangwang replied:
ack.

hachikuji commented on core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala:
```diff
@@ -109,73 +109,91 @@ class TransactionCoordinator(brokerId: Int,
+
+      metadata.prepareTransitionTo(Empty)
+
+      initPidWithNewMetadata(transactionalId, responseCallback, epochAndMetadata)
```
Shouldn't we be using `newMetadata` somehow?

guozhangwang replied:
done.

hachikuji commented on core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala:
```diff
@@ -47,19 +47,19 @@ object TransactionCoordinator {
+    new TransactionCoordinator(config.brokerId, pidManager, txnStateManager, txnMarkerChannelManager, txnMarkerPurgatory, time)
   }
 
   private def initTransactionError(error: Errors): InitPidResult = {
```
nit: Not from this patch, but couldn't we add a couple constructors to `InitPidResult` and remove these functions? Or maybe just add default values for PID and epoch.

guozhangwang replied:
I have also thought about that, e.g. having sth. like

```
case class InitPidResult(pid: Long = 
RecordBatch.NO_PRODUCER_ID, epoch: Short = RecordBatch.NO_PRODUCER_EPOCH, error: Errors)
```

and call it without overriding defaults upon checking failure cases, but I found it is less readable. If you feel strong about it I can change.

hachikuji replied:
No strong preference if you already rejected the option.

hachikuji commented on core/src/main/scala/kafka/coordinator/transaction/TransactionMetadata.scala:
```diff
@@ -138,37 +138,115 @@ private[coordinator] class TransactionMetadata(val pid: Long,
   }
 
-  def completeTransitionTo(newState: TransactionState): Boolean = {
+  def completeTransitionTo(newMetadata: TransactionMetadata): Boolean = {
```
The reuse of `TransactionMetadata` here is super confusing. It seems we just need a struct to propagate the new state to the completion callback. Maybe we could do this with an immutable case class instead? Maybe `TransactionStateTransition` or something.

guozhangwang replied:
Not sure I understand this comment, could you elaborate a bit?

guozhangwang replied:
Discussed offline.

hachikuji commented on core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerChannelManager.scala:
```diff
@@ -98,55 +99,91 @@ object TransactionMarkerChannelManager {
+    def appendToLogCallback(error: Errors): Unit = {
+      error match {
+        case Errors.NONE =>
+          trace(s"Competed sending transaction markers for $transactionalId as $txnResult")
```
typo: "Competed"

guozhangwang replied:
ack.

hachikuji commented on core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala:
```diff
@@ -109,73 +109,91 @@ class TransactionCoordinator(brokerId: Int,
+    // after it was successfully written to log
+    metadata.prepareTransitionTo(Empty)
+
+    txnManager.appendTransactionToLog(transactionalId, epochAndMetadata.coordinatorEpoch, epochAndMetadata.transactionMetadata, callback)
```
This is pretty confusing stuff. My understanding is that once this append completes, we'll invoke `TransationMetadata.completeTransitionTo`. In this case, it will be the same `TransactionMetadata` object passed to itself. Might not be incorrect, but it's definitely weird.

By the way, you can replace `epochAndMetadata.transactionMetadata` with `metadata`.

guozhangwang replied:
I just refactored the code a bit around `initPid`, since it is a special case compared with others:

1. In other cases, we first create a new metadata as a place holder of all the pending updates to the original object and then only update it in the callback after append / send marker completes;

2. In `initPid` case, we would try first inserting a dummy into the cache but set its pending state, as `Empty -> Empty`; if there is an existing entry already we would either a) abort its existing txn first b) return retriable error code or c) update epoch / txn timeout and do the same as case 1) above.

The new code path looks better to me know. Let me know what do you think.

hachikuji reviewed (commented):
Left a few comments. I'm still working my way through the whole thing (been out of the TC code for too long).

ijuma commented on core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala:
```diff
@@ -194,35 +212,37 @@ class TransactionCoordinator(brokerId: Int,
       // try to update the transaction metadata and append the updated metadata to txn log;
       // if there is no such metadata treat it as invalid pid mapping error.
-      val (error, newMetadata) = txnManager.getTransactionState(transactionalId) match {
+      val (error: Errors, coordinatorEpoch: Int, txnMetadata: TransactionMetadata) = txnManager.getTransactionState(transactionalId) match {
```
Similar to a comment that I left in a different PR, adding type annotations here triggers a pattern match and it's brittle if there are cases where the match can fail, but the compiler can't prove it (any time there is a subclass involved). It's one of the cases where relying on type inference is safer (it will always infer a safe type).

ijuma replied:
Btw, a better way to represent this would be using `Either[Errors, (Int, TransactionMetadata)]`. Even better would be to have a type for the tuple on the right, but even without it, it would make the flow way clearer IMO.

guozhangwang replied:
ack.

guozhangwang replied:
EDIT: @ijuma I realized that in many cases I would return a `null`, and it will cause the annotated type to be `Any` at compilation time. So I think it is actually better to enforce type annotations in this case.

ijuma replied:
Yeah, you should never use `null` in Scala generally. Using an `Either` here avoids all the issues.

guozhangwang replied:
Got it, thanks!

junrao commented on core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala:
```diff
@@ -374,37 +372,33 @@ class TransactionStateManager(brokerId: Int,
+            epochAndMetadata synchronized {
+              val metadata = epochAndMetadata.transactionMetadata
+
+              if (epochAndMetadata.coordinatorEpoch == coordinatorEpoch && metadata.completeTransitionTo(txnMetadata)) {
```
Not sure checking coordinatorEpoch here is enough since coordinatorEpoch could change when we append the log.

guozhangwang replied:
Not sure I understand your question? `coordinatorEpoch` is the value passed from the caller, which is the epoch of the cached metadata object before append is called, and here we are checking if it is still the same epoch by reading from the cache again. Plus inside `completeTransitionTo` we also check that other fields inside metadata are expected.

guozhangwang committed:
github comments

guozhangwang committed:
more github comments

junrao commented on core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala:
```diff
@@ -106,75 +106,97 @@ class TransactionCoordinator(brokerId: Int,
+            // with the transactional id at the same time (hence reference equality will fail);
+            // in this case we will treat it as the metadata has existed already
+            txnMetadata synchronized {
+              if (txnMetadata != createdMetadata) {
```
In scala, == tests object equality. I think we want to test reference equality here. If so, we should use eq for testing.

guozhangwang replied:
ack.

junrao commented on core/src/main/scala/kafka/coordinator/transaction/TransactionMetadata.scala:
```diff
@@ -125,50 +134,167 @@ private[coordinator] class TransactionMetadata(val pid: Long,
 
-  def prepareTransitionTo(newState: TransactionState): Boolean = {
+  def prepareNoTransit(): TransactionMetadataTransition =
+    prepareTransitionTo(state, producerEpoch.toShort, txnTimeoutMs, immutable.Set.empty[TopicPartition], txnStartTimestamp, txnLastUpdateTimestamp)
```
Hmm, this means that in prepareTransitionTo(), we need to transition from one state to itself. Not sure if all states are allowed to transition from itself.

guozhangwang replied:
We only allow `Empty` -> `Empty` transition, and it is a special case for adding a txnID for the first time, the inserted entry will be Empty and its pending state also Empty. Only when the pending state is cleared after the txn log write returns it is considered "created succesfully".

As for this function itself, it is only used when `initPid` is received while the current PID has an ongoing txn, i.e. it is `Ongoing`, and later we check if it is `Ongoing` we will abort the txn first and return `CONCURRENT_TXN` to let user retry.

junrao commented on core/src/main/scala/kafka/coordinator/transaction/TransactionMetadata.scala:
```diff
@@ -125,50 +134,167 @@ private[coordinator] class TransactionMetadata(val pid: Long,
+  def prepareNoTransit(): TransactionMetadataTransition =
+    prepareTransitionTo(state, producerEpoch.toShort, txnTimeoutMs, immutable.Set.empty[TopicPartition], txnStartTimestamp, txnLastUpdateTimestamp)
+
+  def prepareIncrementEpoch(newTxnTimeoutMs: Int,
```
Since now there are different types of epoch, would it be better to name this prepareIncrementProducerEpoch?

guozhangwang replied:
ack.

junrao commented on core/src/main/scala/kafka/coordinator/transaction/TransactionMetadata.scala:
```diff
@@ -125,50 +134,167 @@ private[coordinator] class TransactionMetadata(val pid: Long,
-
   override def hashCode(): Int = {
-    val state = Seq(pid, txnTimeoutMs, topicPartitions, lastUpdateTimestamp)
+    val state = Seq(producerId, txnTimeoutMs, topicPartitions, txnLastUpdateTimestamp)
```
Should we include pendingState in hashCode() and equals()?

guozhangwang replied:
ack.

junrao commented on core/src/main/scala/kafka/coordinator/transaction/TransactionMetadata.scala:
```diff
@@ -125,50 +134,167 @@ private[coordinator] class TransactionMetadata(val pid: Long,
+          if (producerEpoch != newMetadata.producerEpoch ||
+            !topicPartitions.subsetOf(newMetadata.topicPartitions) ||
+            txnTimeoutMs != newMetadata.txnTimeoutMs ||
+            txnStartTimestamp < newMetadata.txnStartTimestamp) {
```
Hmm, when transitioning from empty to ongoing, it seems it's ok for newMetadata.txnStartTimestamp to be larger than current txnStartTimestamp. It's only when transitioning from ongoing to ongoing that we don't want txnStartTimestamp to change.

guozhangwang replied:
That is true: when the pid is first created, there is no txn yet, the startTime is set to "-1", when we received the first addPartitions we will update the startTime to `now`, and it will not be updated until it has completed, and then a new addPartition is received indicating the start of a new txn for this pid.

junrao commented on core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala:
```diff
@@ -106,75 +106,97 @@ class TransactionCoordinator(brokerId: Int,
+      // caller should have synchronized on txnMetadata already
+      txnMetadata.state match {
+        case PrepareAbort | PrepareCommit =>
+          // reply to client and let it retry immediately
```
It seems the comment should be the same as in line 185: let client backoff and rety instead of retry immediately.

guozhangwang replied:
ack.

junrao commented on core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala:
```diff
@@ -149,118 +160,112 @@ class TransactionStateManager(brokerId: Int,
   }
 
-  private def loadTransactionMetadata(topicPartition: TopicPartition, writeTxnMarkers: WriteTxnMarkers) {
+  private def loadTransactionMetadata(topicPartition: TopicPartition, coordinatorEpoch: Int): Pool[String, TransactionMetadata] =  {
```
highWaterMark below should be renamed to logEndOffset?

guozhangwang replied:
ack.

junrao commented on core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala:
```diff
@@ -271,31 +276,23 @@ class TransactionStateManager(brokerId: Int,
 
-        if (numTxnsRemoved > 0)
-          info(s"Removed $numTxnsRemoved cached transaction metadata for $topicPartition on follower transition")
+        loadingPartitions.remove(partitionId)
```
Since removeTransactions() is cheap, not sure if this needs to be run in a scheduler. Also, Perhaps it's useful to wait for the scheduler to have no pending task here before return? Any pending loadTransaction task should be completed very quickly after removeTransactions() is called.

guozhangwang replied:
> Any pending loadTransaction task should be completed very quickly after removeTransactions() is called.

We could be loading for partition X while removing for partition Y in this case, so waiting for the removing to complete may still take long time right?

Another benefit to let all three operations (loading, removing-partition, expiring-txns) to be execute by the same back ground thread is that locking mechanism would be a bit easier to reason (this is not part of this PR): only background thread would grab the write lock, and the handler thread will only try to grab read lock, and handler thread would not add / remove any entries from the maps but just modify the objects in-place.

junrao commented on core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerChannel.scala:
```diff
@@ -33,48 +33,41 @@ import scala.collection.{concurrent, immutable, mutable}
 import collection.JavaConverters._
 
 class TransactionMarkerChannel(interBrokerListenerName: ListenerName,
+                               txnStateManager: TransactionStateManager,
```
Does BrokerRequestQueue.destination need to be volatile?

guozhangwang replied:
Good point! ack.

junrao commented on core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerChannel.scala:
```diff
@@ -84,102 +77,70 @@ class TransactionMarkerChannel(interBrokerListenerName: ListenerName,
-            trace(s"couldn't find leader for partition: $topicPartition")
-        }
-        if (brokerId.isEmpty)
+      while (brokerNode.isEmpty) {
```
Hmm, the while loop may tie up a request handler thread, which is not ideal.

guozhangwang replied:
yeah I have thought about that too.. The approach is that if we cannot locate the broker to offer to its queue then in order for this thread to return we likely need to put it in an "unknown broker" queue first, and periodically we can check the queue and migrate to the found brokers. Does that make sense?

junrao commented on core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerChannelManager.scala:
```diff
@@ -98,55 +99,92 @@ object TransactionMarkerChannelManager {
+    // watch for both the transactional id and the transaction topic partition id,
+    // so we can cancel all the delayed operations for the same partition id;
+    // NOTE this is only possible because the hashcode of Int / String never overlaps
+    val delayedTxnMarker = new DelayedTxnMarker(txnMetadata, appendToLogCallback)
```
Since the timeout for delayedTxnMarker is infinite, do we need a purgatory or just a map?

guozhangwang replied:
I have not made up my mind about infinite timeout, and would like to have another discussion with you (as I left some comments in Damian's previous patch: 

https://github.com/apache/kafka/pull/2849#discussion_r113353056

For now I'll add a TODO marker and do that in a follow-up once we have a conclusion.

junrao commented on core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerRequestCompletionHandler.scala:
```diff
@@ -22,25 +22,25 @@ import kafka.utils.Logging
-      for (txnMarker: TxnMarkerEntry <- txnMarkerEntries) {
-        transactionMarkerChannel.addRequestToSend(
-          txnTopicPartition,
+      for (txnIdAndMarker: TxnIdAndMarkerEntry <- txnIdAndMarkerEntries) {
```
Hmm, in the disconnected case, shouldn't we check txnStateManager.getTransactionState as well in case an emmigration has happened?

guozhangwang replied:
We check that in the `appendToLogCallback` callback in `txnMarkerChannelManager`. Generally speaking here is the reasoning:

1. for the handler logic, the only error code that we should not retry is "coordinatorEpochNotValid", in which we should complete the delayed operation immediately with a meaningful error code (currently it is always `NONE` ). 

2. then in `appendToLogCallback`, we check the error code, and if it is due to received `coordinatorEpochNotValid` we can simply ignore the rest of the operations.

junrao replied:
Yes, it's just that in the case that a transaction coordinator's epoch has changed, there is no need to keep resending the WriteMarker request to the brokers.

guozhangwang replied:
Makes sense. I will add the check.

junrao commented on core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerRequestCompletionHandler.scala:
```diff
@@ -52,43 +52,50 @@ class TransactionMarkerRequestCompletionHandler(transactionMarkerChannel: Transa
 
-        if (errors == null)
-          throw new IllegalStateException("WriteTxnMarkerResponse does not contain expected error map for pid " + txnMarker.producerId())
+        txnStateManager.getTransactionState(transactionalId) match {
```
Since multiple immigration/emigration could have happened when the response comes back, a more reliable way is to check if the current coordinator epoch is still the same as what's in the request.

guozhangwang replied:
ack. I have re-factored the handling logic here to cover all the error codes, please take a look at the modified file.

junrao commented on core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerRequestCompletionHandler.scala:
```diff
@@ -52,43 +52,50 @@ class TransactionMarkerRequestCompletionHandler(transactionMarkerChannel: Transa
+        txnStateManager.getTransactionState(transactionalId) match {
+          case None =>
+            // txn topic partition has likely emigrated, just cancel it from the purgatory
+            txnMarkerPurgatory.cancelForKey(transactionalId)
```
Should we also remove the entries keyed by transactionalId in purgatory?

guozhangwang replied:
Not sure I understand the comment, we are indeed canceling for the transactionalId right?

junrao commented on core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerRequestCompletionHandler.scala:
```diff
@@ -52,43 +52,50 @@ class TransactionMarkerRequestCompletionHandler(transactionMarkerChannel: Transa
+              for ((topicPartition: TopicPartition, error: Errors) <- errors) {
+                error match {
+                  case Errors.NONE =>
+                    epochAndMetadata.transactionMetadata.topicPartitions -= topicPartition
```
Structure wise, maybe it's better to do this as part of TransactionMetadata.completeTransitionTo() so that we can limit the place where internals of TransactionMetadata are modified?

hachikuji replied:
Also it seems like a good idea to verify that the state of the transaction is still what we expect.

junrao commented on core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerRequestCompletionHandler.scala:
```diff
@@ -52,43 +52,50 @@ class TransactionMarkerRequestCompletionHandler(transactionMarkerChannel: Transa
+                txnMarker.coordinatorEpoch(),
+                retryPartitions.toSet)
+            } else {
+              txnMarkerPurgatory.checkAndComplete(txnMarker.producerId)
```
Hmm, is this right? We watch on transactionalId, not producerId. Also, since txnMarkerPurgatory is passed around to different classes, it's bit hard to track who is calling checkAndComplete on this purgatory. An alternative way is to expose an access method to txnMarkerPurgatory in TransactionChannelManager and let all classes call that method.

guozhangwang replied:
ack.

junrao reviewed (commented):
@guozhangwang : Thanks for the patch. Left a few comments.

hachikuji commented on core/src/main/scala/kafka/coordinator/transaction/TransactionMetadata.scala:
```diff
@@ -125,50 +134,167 @@ private[coordinator] class TransactionMetadata(val pid: Long,
 
-  def prepareTransitionTo(newState: TransactionState): Boolean = {
+  def prepareNoTransit(): TransactionMetadataTransition =
+    prepareTransitionTo(state, producerEpoch.toShort, txnTimeoutMs, immutable.Set.empty[TopicPartition], txnStartTimestamp, txnLastUpdateTimestamp)
```
nit: the `toShort` is not needed.

guozhangwang replied:
ack.

hachikuji commented on core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala:
```diff
@@ -106,75 +106,97 @@ class TransactionCoordinator(brokerId: Int,
+          // try to append and then update
+          Right(coordinatorEpoch, txnMetadata.prepareIncrementEpoch(transactionTimeoutMs, time.milliseconds()))
+
+        case _ =>
```
Maybe we could use the `Ongoing` state explicitly?

guozhangwang replied:
ack.

hachikuji commented on core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala:
```diff
@@ -106,75 +106,97 @@ class TransactionCoordinator(brokerId: Int,
       } else {
         // only try to get a new pid and update the cache if the transactional id is unknown
-        txnManager.getTransactionState(transactionalId) match {
+        val result: Either[InitPidResult, (Int, TransactionMetadataTransition)] = txnManager.getTransactionState(transactionalId) match {
```
nit: the whole body of `handleInitPid` is misaligned.

guozhangwang replied:
ack.

hachikuji commented on core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala:
```diff
@@ -194,189 +216,149 @@ class TransactionCoordinator(brokerId: Int,
-              (Errors.CONCURRENT_TRANSACTIONS, null)
+              Left(Errors.CONCURRENT_TRANSACTIONS)
+            } else if (txnMetadata.state == PrepareCommit || txnMetadata.state == PrepareAbort) {
+              Left(Errors.CONCURRENT_TRANSACTIONS)
```
Hmm.. If the user is trying to add partitions before the previous transaction has completed, shouldn't that be INVALID_TXN_STATE?

guozhangwang replied:
Well, since we return to the user right after we have appended the log, in which case the state has updated to `PrepareXX`, the client code is possible to send another `addPartitions` while the sending markers are still on the flight. In this case we should not return a fatal error but let client retry.

hachikuji replied:
Ack, makes sense.

hachikuji commented on core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala:
```diff
@@ -194,189 +216,149 @@ class TransactionCoordinator(brokerId: Int,
-                if (command == TransactionResult.COMMIT)
-                  responseCallback(Errors.NONE)
+                if (txnMarkerResult == TransactionResult.COMMIT)
+                  // this is an optimization: return OK immediately
```
nit: I think this is more than an optimization: it is necessary for correctness because there is no guarantee that the client will receive the result of the EndTxnRequest.

guozhangwang replied:
Okay, will update the comments.

hachikuji commented on core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala:
```diff
@@ -194,189 +216,149 @@ class TransactionCoordinator(brokerId: Int,
-                if (command == TransactionResult.ABORT)
-                  responseCallback(Errors.NONE)
+                if (txnMarkerResult == TransactionResult.ABORT)
+                // this is an optimization: return OK immediately
```
nit: comment misaligned. Also, pity we can't merge this branch with `CompleteCommit` somehow.

guozhangwang replied:
ack.

hachikuji commented on core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala:
```diff
@@ -194,189 +216,149 @@ class TransactionCoordinator(brokerId: Int,
                 else
-                  responseCallback(Errors.INVALID_TXN_STATE)
+                  Left(Errors.INVALID_TXN_STATE)
               case _ =>
```
Hmm... Isn't it possible that a client resends an EndTxnRequest while we are still in `PrepareCommit` or `PrepareAbort`. As long as the outcome matches, it seems we should accept those requests and perhaps return CONCURRENT_TRANSACTIONS?

Also, can we list the states we're catching here explicitly?

guozhangwang replied:
Good point! ack.

guozhangwang replied:
EDIT: actually I think for most cases this case should be covered in `txnMetadata.pendingTransitionInProgress`, the only edge case that we have migrated to `prepareXX` but have not prepareTo `completeXX` (note they should happen fairly consecutively). Anyways, will add them to cover it as well.

hachikuji commented on core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala:
```diff
@@ -194,189 +216,149 @@ class TransactionCoordinator(brokerId: Int,
+            // the log append was successful
+            responseCallback(error)
+
+            if (error == Errors.NONE) {
```
Maybe we should log something in the else case?

guozhangwang replied:
ack.

hachikuji commented on core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala:
```diff
@@ -194,189 +216,149 @@ class TransactionCoordinator(brokerId: Int,
+                  sendTxnMarkers(transactionalId, coordinatorEpoch, metadata, txnMarkerResult)
+
+                case None =>
+                  // this transactional id no longer exists, maybe the corresponding partition has already been migrated out.
```
Could we check for the case explicitly with `txnManager.isCoordinatorFor(transactionalId)`? If the transactionalId did not migrate, then maybe it should be an illegal state.

guozhangwang replied:
good point, ack.

hachikuji commented on core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala:
```diff
@@ -410,10 +392,4 @@ class TransactionCoordinator(brokerId: Int,
-                              nextState: TransactionState,
-                              newMetadata: TransactionMetadata,
-                              coordinatorEpoch: Int)
+case class InitPidResult(pid: Long, epoch: Short, error: Errors)
```
nit: add newline

guozhangwang replied:
ack.

hachikuji reviewed (commented):
A few comments in `TransactionCoordinator`.

hachikuji commented on core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerRequestCompletionHandler.scala:
```diff
@@ -52,43 +52,50 @@ class TransactionMarkerRequestCompletionHandler(transactionMarkerChannel: Transa
+            if (errors == null)
+              throw new IllegalStateException(s"WriteTxnMarkerResponse does not contain expected error map for pid ${txnMarker.producerId}")
+
+            epochAndMetadata synchronized {
```
This should be `epochAndMetadata.transactionMetadata`?

guozhangwang replied:
ack. This will be part of the minor refactoring that exhausts all the possible error codes in the writeTxnMarker responses.

hachikuji commented on core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerChannelManager.scala:
```diff
@@ -98,55 +99,92 @@ object TransactionMarkerChannelManager {
+
+          txnStateManager.getTransactionState(transactionalId) match {
+            case Some(epochAndMetadata) =>
+              epochAndMetadata synchronized {
```
Same here: `epochAndMetadata.transactionMetadata`.

guozhangwang replied:
Actually I just realized we do not need to check the coordinator epoch anymore, since it is already checked in `appendTransactionToLog#updateCacheCallback`, which is executed before calling this callback, so the error code returned is already reflecting the fact if 

```
(epochAndMetadata.coordinatorEpoch == coordinatorEpoch && metadata.completeTransitionTo(newMetadata))
```

So we can directly go head with the error code here.

hachikuji commented on core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerChannel.scala:
```diff
@@ -33,48 +33,41 @@ import scala.collection.{concurrent, immutable, mutable}
-  // we need the txnTopicPartition so we can clean up when Transaction Log partitions emigrate
-  case class PendingTxnKey(txnTopicPartition: Int, producerId: Long)
-
   class BrokerRequestQueue(private var destination: Node) {
```
Seems this class doesn't depend on the containing instance of `TransactionMarkerChannel`. Can we move it outside or to the companion object?

guozhangwang replied:
Ack.

hachikuji commented on core/src/main/scala/kafka/coordinator/transaction/TransactionMetadata.scala:
```diff
@@ -89,32 +89,41 @@ private[coordinator] object TransactionMetadata {
 }
 
+// this is a immutable object representing the target transition of the transaction metadata
+private[coordinator] case class TransactionMetadataTransition(producerId: Long,
```
nit: `private[transaction]`?

guozhangwang replied:
ack.

hachikuji commented on core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerChannel.scala:
```diff
@@ -33,48 +33,41 @@ import scala.collection.{concurrent, immutable, mutable}
-      requestsPerTxnTopicPartition.map { case(_, queue) => queue.size()}
-        .sum
-
+    def totalQueuedRequests(): Int = markersPerTxnTopicPartition.map { case(_, queue) => queue.size()}.sum
```
Might be worth noting that `size` is O(n). Also, maybe `Queued` is redundant given the name of the class. Maybe it could just be `totalRequests`?

guozhangwang replied:
Ack.

hachikuji commented on core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerChannel.scala:
```diff
@@ -33,48 +33,41 @@ import scala.collection.{concurrent, immutable, mutable}
 import collection.JavaConverters._
 
 class TransactionMarkerChannel(interBrokerListenerName: ListenerName,
```
Could this be `private[transaction]`? Also I found the name a bit misleading: would `TransactionMarkerQueue` or `TransactionMarkerAccumulator` be closer?

guozhangwang replied:
ack. After thinking about this, I'm going to merge the `markerChannel` into `markerChannelManager` since it is always a one-to-one mapping.

hachikuji commented on core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerChannel.scala:
```diff
@@ -33,48 +33,41 @@ import scala.collection.{concurrent, immutable, mutable}
+    def totalQueuedRequests(): Int = markersPerTxnTopicPartition.map { case(_, queue) => queue.size()}.sum
   }
 
   private val brokerStateMap: concurrent.Map[Int, BrokerRequestQueue] = concurrent.TrieMap.empty[Int, BrokerRequestQueue]
```
Maybe `brokerRequestQueues`?

guozhangwang replied:
ack.

hachikuji commented on core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerChannelManager.scala:
```diff
@@ -71,18 +71,19 @@ object TransactionMarkerChannelManager {
+    val txnMarkerSendThread: InterBrokerSendThread = {
       networkClient.wakeup()
-      new InterBrokerSendThread(threadName, networkClient, requestGenerator(channel, txnMarkerPurgatory), time)
+      new InterBrokerSendThread(threadName, networkClient, requestGenerator(txnMarkerChannel, txnMarkerPurgatory), time)
```
It would be nice to decouple `TransactionStateManager` from `TransactionMarkerChannel`. One way to do this is to move the building of the `RequestAndCompletionHandler` objects into `requestGenerator`. You can let `drainQueuedTransactionMarkers` return something like `Map[Int, List[TxnIdAndMarkerEntry]]`. Then `TransactionMarkerChannel` would not need a reference to `TransactionStateManager`. Also, the generator pattern is a little odd. Could we just make `InterBrokerSendThread` an abstract class with an abstract method `generateRequests`? In that case, you could let `TransactionMarkerChannelManager` extend directly from `InterBrokerSendThread` and implement the request generation.

hachikuji commented on core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerChannel.scala:
```diff
@@ -84,102 +77,70 @@ class TransactionMarkerChannel(interBrokerListenerName: ListenerName,
-
-  def pendingTxnMetadata(metadataPartition: Int, pid: Long): Option[TransactionMetadata] = {
-    pendingTxnMap.get(PendingTxnKey(metadataPartition, pid))
+    networkClient.wakeup()
```
Kind of unfortunate that we have a dependence on `NetworkClient` only in order to invoke this `wakeup()`. I'm wondering if the callers could do it instead and we can remove the dependence? That will help to decouple the objects which will make testing easier.

guozhangwang replied:
ack. Let me know WDYT about the after-refactoring.

hachikuji commented on core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerChannelManager.scala:
```diff
@@ -71,18 +71,19 @@ object TransactionMarkerChannelManager {
 
-    val sendThread: InterBrokerSendThread = {
+    val txnMarkerSendThread: InterBrokerSendThread = {
       networkClient.wakeup()
```
Why do we do this? Maybe worth a comment.

guozhangwang replied:
I'm not sure either, it is from the old code. My understanding is that `networkClient` is only pollable from the `sendThread`, and hence `wakeup` is not required. But maybe I'm missing something here, cc @dguy

hachikuji commented on core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerRequestCompletionHandler.scala:
```diff
@@ -52,43 +52,50 @@ class TransactionMarkerRequestCompletionHandler(transactionMarkerChannel: Transa
+                    epochAndMetadata.transactionMetadata.topicPartitions -= topicPartition
+
+                  case Errors.UNKNOWN_TOPIC_OR_PARTITION | Errors.NOT_LEADER_FOR_PARTITION |
+                       Errors.NOT_ENOUGH_REPLICAS | Errors.NOT_ENOUGH_REPLICAS_AFTER_APPEND =>
```
I think `REQUEST_TIMEOUT` might be another possibility.

guozhangwang replied:
I checked the source code, `TimeoutException` can only be thrown from producer / consumer internals, but they will never be returned from the broker side. So this should not be a possibility? @hachikuji

hachikuji replied:
@guozhangwang Did you also look for uses of `Errors.REQUEST_TIMEOUT`?

guozhangwang replied:
You mean `REQUEST_TIMED_OUT`? There is no `REQUEST_TIMEOUT` in Errors. For the latter case I searched the code and found no return errors from broker-side.

hachikuji commented on core/src/main/scala/kafka/coordinator/transaction/TransactionMetadata.scala:
```diff
@@ -125,50 +134,167 @@ private[coordinator] class TransactionMetadata(val pid: Long,
+      producerId != newMetadata.producerId ||
+      txnLastUpdateTimestamp <= newMetadata.txnLastUpdateTimestamp) {
+
       false
```
I wonder if you can add a comment explaining when this case can happen.

hachikuji commented on core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala:
```diff
@@ -149,118 +160,112 @@ class TransactionStateManager(brokerId: Int,
+            && !shuttingDown.get()) {
+            buffer.clear()
+            val fileRecords = log.read(currOffset, config.transactionLogLoadBufferSize, maxOffset = None, minOneMessage = true)
+              .records.asInstanceOf[FileRecords]
```
The cast to `FileRecords` is not totally safe. I fixed the same problem in `GroupMetadataManager` recently.

guozhangwang replied:
rebased this already.

hachikuji commented on core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala:
```diff
@@ -149,118 +160,112 @@ class TransactionStateManager(brokerId: Int,
+    val txnMetadataCacheEntry = TxnMetadataCacheEntry(coordinatorEpoch, metadataPerTransactionalId)
+    val currentTxnMetadataCacheEntry = transactionMetadataCache.put(txnTopicPartition, txnMetadataCacheEntry)
+
+    if (currentTxnMetadataCacheEntry.isDefined) {
```
nit: pattern match?

guozhangwang replied:
Could you elaborate a bit? We only have logic for the `if` condition?

hachikuji commented on core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala:
```diff
@@ -345,73 +343,69 @@ class TransactionStateManager(brokerId: Int,
+              if (epochAndMetadata.coordinatorEpoch == coordinatorEpoch && metadata.completeTransitionTo(newMetadata)) {
+                debug(s"Updating $transactionalId's transaction state to $newMetadata with coordinator epoch $coordinatorEpoch for $transactionalId succeeded")
+              } else {
+                // the cache may have been changed due to txn topic partition emigration and immigration,
```
A cache change due to emigration/immigration would be handled by the epoch check, right? Are there any cases where `completeTransitionTo` itself could fail?

guozhangwang replied:
Good point, we could consider throw exception directly instead of returning false.

@junrao @hachikuji There are a few of those suggestions that we can do in a follow-up PR. I have marked those places with `TODO`s.

guozhangwang committed:
rebase from trunk

guozhangwang committed:
Jun's comments

guozhangwang committed:
Jason's comments

guozhangwang committed:
Merge branch 'trunk' of https://git-wip-us.apache.org/repos/asf/kafka into K5130-refactor-tc-inmemory-cache

guozhangwang committed:
a few more minor fixes

guozhangwang committed:
list all possible exception error code in txn marker request completion handler

guozhangwang committed:
fix unit tests

guozhangwang commented:
@hachikuji @dguy @junrao I have addressed your comments. Two major modifications:

1. `TransactionMarkerRequestCompletionHandler`: exhaust all possible error codes.
2. Merge `TransactionMarkerChannel` into `TransactionMarkerChannelManager`.

guozhangwang committed:
fix more unit tests

guozhangwang committed:
fix more unit tests

guozhangwang committed:
add txnIdAndMarkersEntry

hachikuji commented on core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerChannelManager.scala:
```diff
@@ -71,79 +75,211 @@ object TransactionMarkerChannelManager {
+  def totalNumMarkers(): Int = markersPerTxnTopicPartition.map { case(_, queue) => queue.size()}.sum
+}
 
 class TransactionMarkerChannelManager(config: KafkaConfig,
```
After we remove `TransactionMarkerChannel`, maybe we can claim its name for this class?

hachikuji commented on core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerRequestCompletionHandler.scala:
```diff
@@ -52,43 +51,84 @@ class TransactionMarkerRequestCompletionHandler(transactionMarkerChannel: Transa
-                  metadata.topicPartitions -= topicPartition
+        txnStateManager.getTransactionState(transactionalId) match {
+          case None =>
+            info(s"Transaction topic partition for $transactionalId may likely has emigrated, as the corresponding metadata do not exist in the cache" +
```
Can we mention the partition number? Also, "may likely has emigrated" -> "has likely emigrated"?

hachikuji commented on core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerRequestCompletionHandler.scala:
```diff
@@ -52,43 +51,84 @@ class TransactionMarkerRequestCompletionHandler(transactionMarkerChannel: Transa
+              info(s"Transaction coordinator epoch for $transactionalId has changed from ${txnMarker.coordinatorEpoch} to " +
+                s"${epochAndMetadata.coordinatorEpoch}; cancel sending transaction markers $txnMarker to the brokers")
+
+              txnMarkerChannelManager.removeMarkersForTxnId(transactionalId)
```
If the epoch has changed, is it still necessary or safe to continue with the logic below? In particular, we could still remove the partitions from `TransactionMetadata`. Maybe we should just return?

guozhangwang replied:
This call is just for removing the delayed operation in the txn marker purgatory. I think it is still safe.

hachikuji commented on core/src/main/scala/kafka/coordinator/transaction/TransactionMetadata.scala:
```diff
@@ -125,50 +134,177 @@ private[coordinator] class TransactionMetadata(val pid: Long,
 
-  def completeTransitionTo(newState: TransactionState): Boolean = {
+  // TODO: instead of returning false, we could directly throw exception since it should never be expected
+  def completeTransitionTo(newMetadata: TransactionMetadataTransition): Boolean = {
```
We discussed offline changing this to assert a valid transition instead of returning a boolean. We could do this in a follow-up if you prefer.

guozhangwang replied:
It is done already, but was not pushed somehow, could you check again now?

hachikuji commented on core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerRequestCompletionHandler.scala:
```diff
@@ -52,43 +51,84 @@ class TransactionMarkerRequestCompletionHandler(transactionMarkerChannel: Transa
+                    info(s"Sending $transactionalId's transaction marker for partition $topicPartition has permanently failed with error ${error.exceptionName} " +
+                      s"with the current coordinator epoch ${epochAndMetadata.coordinatorEpoch}; cancel sending any more transaction markers $txnMarker to the brokers")
+
+                    txnMarkerChannelManager.removeMarkersForTxnId(transactionalId)
```
Should we return here? Otherwise, the call to `completeSendMarkersForTxnId` will be invoked below. Not sure it's a problem, but seems odd.

Also, I'm not sure I fully understand the chain of operations that this should trigger. It seems that `removeMarkersForTxnId` cancels the `DelayedTxnMarker`, which means its callback won't get invoked. But what happens to the state of the `TransactionMetadata`? It seems like it will just remain indefinitely in one of the Prepare states. For CoordinatorFenced, that seems OK; we rely on the partition to ultimately be evicted. How about ProducerFenced? I'm actually having a tough time imagining a scenario where we would hit that error. The only one that comes to mind is if the coordinator has become a zombie, in which case we should get blocked by the coordinator epoch. Are there any others? In any case, I think it would be helpful to add some comments here clarifying the scenario and the handling expectation.

guozhangwang replied:
`TransactionMetadata` should be removed by the emmigration handler thread, and in the future the PID expiration scheduler's thread. Other handling logic should never remove the entry, but just to update the entry in-place.

And about returning here: good catch! yeah we should not call `completeSendMarkersForTxnId` in this case. Will fix now.

guozhangwang committed:
rebase and fix unit tests again

guozhangwang committed:
group requests from different txn topic partitions but to the same broker into one request

guozhangwang commented:
@hachikuji addressed your comments, and fixed unit tests.

Also another observation is that we are unnecessarily sending multiple txn marker requests to a single broker for txnIds from different txn topic partitions, I have grouped them into a single request as well in this PR. cc @dguy

guozhangwang committed:
address Jason's comments, round three

guozhangwang committed:
findBug fixes

hachikuji reviewed (approved):
Thanks @guozhangwang, LGTM.

guozhangwang committed:
rebase on trunk

junrao commented on core/src/main/scala/kafka/server/DelayedOperation.scala:
```diff
@@ -118,7 +118,8 @@ object DelayedOperationPurgatory {
                                    brokerId: Int = 0,
-                                   purgeInterval: Int = 1000): DelayedOperationPurgatory[T] = {
+                                   purgeInterval: Int = 1000,
+                                   reaperEnabled: Boolean = true): DelayedOperationPurgatory[T] = {
```
reaperEnabled doesn't seem be be used.

guozhangwang replied:
ack.

guozhangwang committed:
fix unit tests

junrao commented on core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala:
```diff
@@ -93,90 +93,113 @@ class TransactionCoordinator(brokerId: Int,
+                responseCallback(initTransactionError(error))
+            }
+
+            txnManager.appendTransactionToLog(transactionalId, coordinatorEpoch, newMetadata, sendPidResponseCallback)
```
Is appending w/o synchronization on coordinator epoch safe? We don't want to write to the log if the coordinator epoch has changed. The most reliable way is probably to hold a read lock on TransactionStateManager.stateLock() while doing the log append.

guozhangwang replied:
Yup, it is going to be done in the locking PR that I'm also working on now. Just trying to keep each PR small to make it easier for reviews.

guozhangwang commented:
Merged to trunk.

junrao commented on core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala:
```diff
@@ -164,162 +185,159 @@ class TransactionStateManager(brokerId: Int,
     inLock(stateLock) {
-      ownedPartitions.put(partition, coordinatorEpoch)
-      loadingPartitions.add(partition)
+      loadingPartitions.add(partitionId)
```
Hmm, not sure if this is completely safe. In removeTransactionsForTxnTopicPartition(), we modify loadingPartitions in the scheduler and here, we modify loadingPartitions in the method directly. This means that if an emmigration is followed immediately by an immigration, the updating of loadingPartitions by emmigration could happen after the updating of loadingPartitions by immigration, which will leave the state incorrect.

guozhangwang replied:
Good catch! ack.

junrao commented on core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala:
```diff
@@ -164,162 +185,159 @@ class TransactionStateManager(brokerId: Int,
     }
 
-    scheduler.schedule(topicPartition.toString, removeTransactions _)
+    scheduler.schedule(s"remove-txns-for-partition-$topicPartition", removeTransactions _)
```
The issue of changing the transaction states in the scheduler is that when this call returns, the new transaction states are not necessarily reflected. So, another request after this call may still see the old states?

guozhangwang replied:
Got it. However, if we just do the removal in the handler thread, then the immigrate-then-emigrate issue may occur. On the other hand, we cannot depend on scheduler do not have pending request since we are periodically schedule txn-expiration and pid-expiration with that scheduler as well.

So here is what I will do: add a removal partitions which will be modified in the handler thread, then get txn metadata will check if the txn partition is in the removal partitions set.

Then in the locking PR, I will add the read-write-lock so that checking will be monitored by the read lock. WDYT?

junrao commented on core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerRequestCompletionHandler.scala:
```diff
@@ -52,43 +50,89 @@ class TransactionMarkerRequestCompletionHandler(transactionMarkerChannel: Transa
+              s"any more; cancel sending transaction markers $txnMarker to the brokers")
+
+            // txn topic partition has likely emigrated, just cancel it from the purgatory
+            txnMarkerChannelManager.removeMarkersForTxnId(transactionalId)
```
txnMarkerPurgatory registers each item under transactionalId and txnTopicPartition. We are only removing the item registered under transactionalId. So, which process will be removing the item registered under txnTopicPartition, especially the reaper thread is not running in txnMarkerPurgatory?

guozhangwang replied:
It is in the `TransactionMarkerChannelManager#removeMarkersForTxnTopicPartition`.

junrao commented on core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerChannelManager.scala:
```diff
@@ -71,79 +75,226 @@ object TransactionMarkerChannelManager {
+                          txnMetadata: TransactionMetadata,
+                          newMetadata: TransactionMetadataTransition): Unit = {
+
+    def appendToLogCallback(error: Errors): Unit = {
```
Since we always send None as error in DelayedTxnMarker, it seems that we can just get rid of error?

guozhangwang replied:
Hmm that is right, but I feel we will probably need to add other possible error codes as we as fixing various error cases under integration tests, so I'd rather keep it as is for now. If after exactly-once has been quite stable and we still do not have any other error code we can remove it. 

But let mw add a TODO for now in case we forgot..

junrao reviewed (commented):
@guozhangwang : Thanks for the patch. Left a few more comments.

guozhangwang commented:
@junrao Thanks for the comments, I will try to incorporate them in a follow-up PR.
